### PR TITLE
SQL select query for CSV/JSON

### PR DESCRIPTION
**Pull request by sinhaashish**
## Description
select * , select column names has been implemented for CSV.
select * is implemented for JSON. 

## Motivation and Context
S3 select for JSON and CSV is implemented

## Regression
Select * , select columns are working for CSV.
Select * is working for JSON.

tested with python

For CSV
```
import os
import boto3
s3 = boto3.resource('s3',
                    endpoint_url='http://127.0.0.1:9000',
                    aws_access_key_id='minio',
                    aws_secret_access_key='minio123',
                    config=Config(signature_version='s3v4'),
                    region_name='us-east-1')
s3_client = s3.meta.client
r = s3_client.select_object_content(
        Bucket='sinha',
        Key='player.csv',
        ExpressionType='SQL',
	Expression = "select * from S3Object AS S  where _3 ='Acy'  ",
      #Expression = "select * from S3Object AS S  where last_name='Acy'  ",
       InputSerialization = {'CSV': {"FieldDelimiter": ",","FileHeaderInfo":"NONE"}},
       #InputSerialization = {'CSV': {"FieldDelimiter": ",","FileHeaderInfo":"USE"}},
        OutputSerialization = {'CSV': {}},
    )

event_stream = r['Payload']
end_event_received = False
with open('output', 'wb') as f:
    # Iterate over events in the event stream as they come
    for event in event_stream:
        # If we received a records event, write the data to a file
        if 'Records' in event:
            data = event['Records']['Payload']
            f.write(data)
        # If we received a progress event, print the details
        elif 'Progress' in event:
            print(event['Progress']['Details'])
        # End event indicates that the request finished successfully
        elif 'End' in event:
            print('Result is complete')
            end_event_received = True
if not end_event_received:
    raise Exception("End event not received, request incomplete.")
```
For JSON
```
import os
import boto3
s3 = boto3.resource('s3',
                    endpoint_url='http://127.0.0.1:9000',
                    aws_access_key_id='minio',
                    aws_secret_access_key='minio123',
                    config=Config(signature_version='s3v4'),
                    region_name='us-east-1')
s3_client = s3.meta.client

	
r = s3_client.select_object_content(
		Bucket='sinha',
		Key='AmazonExample.json',
		ExpressionType='SQL',
		Expression="select * from s3object s where s.name = 'q' limit 5 ",
		InputSerialization = { 'CompressionType': 'NONE','JSON': {'Type': 'DOCUMENT'}},
		OutputSerialization = {'JSON': { 'RecordDelimiter': '\n',}}
)
	


event_stream = r['Payload']
end_event_received = False
with open('output', 'wb') as f:
    # Iterate over events in the event stream as they come
    for event in event_stream:
        # If we received a records event, write the data to a file
        if 'Records' in event:
            data = event['Records']['Payload']
            f.write(data)
        # If we received a progress event, print the details
        elif 'Progress' in event:
            print(event['Progress']['Details'])
        # End event indicates that the request finished successfully
        elif 'End' in event:
            print('Result is complete')
            end_event_received = True
if not end_event_received:
    raise Exception("End event not received, request incomplete.")
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
```
Error running mint automation
+ mkdir -p 6648-9d28700/gopath/src/github.com/minio
+ git clone --quiet https://github.com/minio/minio.git 6648-9d28700/gopath/src/github.com/minio/minio
++ cd 6648-9d28700/gopath/src/github.com/minio/minio
++ git remote add minio https://github.com/minio/minio.git
++ cd 6648-9d28700/gopath/src/github.com/minio/minio
++ git fetch --quiet minio pull/6648/head:pr6648
++ cd 6648-9d28700/gopath/src/github.com/minio/minio
++ git checkout --quiet pr6648
++ minikube ip
+ minikube_ip=192.168.39.64
++ mkdir -p certs
++ cd certs
++ ./cert -host 192.168.39.64
2018/10/17 00:27:49 wrote public.crt
2018/10/17 00:27:49 wrote private.key
+ GOPATH=/mnt/nvme0n1p1/mint-auto/mint-auto/6648-9d28700/gopath
+ make -C 6648-9d28700/gopath/src/github.com/minio/minio --quiet
vendor/github.com/json-iterator/go/config.go:10:2: cannot find package "github.com/modern-go/concurrent" in any of:
	/mnt/nvme0n1p1/mint-auto/mint-auto/6648-9d28700/gopath/src/github.com/minio/minio/vendor/github.com/modern-go/concurrent (vendor tree)
	/mnt/nvme0n1p1/mint-auto/.gimme/versions/go1.10.4.linux.amd64/src/github.com/modern-go/concurrent (from $GOROOT)
	/mnt/nvme0n1p1/mint-auto/mint-auto/6648-9d28700/gopath/src/github.com/modern-go/concurrent (from $GOPATH)
make: *** [build] Error 1

```

**Comment by minio-ops** [comment]
```
Error running mint automation
+ mkdir -p 6648-f3ddbd4/gopath/src/github.com/minio
+ git clone --quiet https://github.com/minio/minio.git 6648-f3ddbd4/gopath/src/github.com/minio/minio
++ cd 6648-f3ddbd4/gopath/src/github.com/minio/minio
++ git remote add minio https://github.com/minio/minio.git
++ cd 6648-f3ddbd4/gopath/src/github.com/minio/minio
++ git fetch --quiet minio pull/6648/head:pr6648
++ cd 6648-f3ddbd4/gopath/src/github.com/minio/minio
++ git checkout --quiet pr6648
++ minikube ip
+ minikube_ip=192.168.39.64
++ mkdir -p certs
++ cd certs
++ ./cert -host 192.168.39.64
2018/10/17 12:50:27 wrote public.crt
2018/10/17 12:50:27 wrote private.key
+ GOPATH=/mnt/nvme0n1p1/mint-auto/mint-auto/6648-f3ddbd4/gopath
+ make -C 6648-f3ddbd4/gopath/src/github.com/minio/minio --quiet
vendor/github.com/json-iterator/go/config.go:10:2: cannot find package "github.com/modern-go/concurrent" in any of:
	/mnt/nvme0n1p1/mint-auto/mint-auto/6648-f3ddbd4/gopath/src/github.com/minio/minio/vendor/github.com/modern-go/concurrent (vendor tree)
	/mnt/nvme0n1p1/mint-auto/.gimme/versions/go1.10.4.linux.amd64/src/github.com/modern-go/concurrent (from $GOROOT)
	/mnt/nvme0n1p1/mint-auto/mint-auto/6648-f3ddbd4/gopath/src/github.com/modern-go/concurrent (from $GOPATH)
make: *** [build] Error 1

```

**Comment by deekoder** [comment]
@sinhaashish will review after resolving conflicts.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by harshavardhana** [comment]
@sinhaashish you also removed the commit I sent to your branch. Please do not rebase and force push your local change without bringing the latest changes to your branch. 

It requires quite a bit of effort to re-work the PR.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by sinhaashish** [comment]
> @sinhaashish you also removed the commit I sent to your branch. Please do not rebase and force push your local change without bringing the latest changes to your branch.
> 
> It requires quite a bit of effort to re-work the PR.

Sorry I pushed force by mistake, I have added the changes manually, PTAL

**Review by harshavardhana — CHANGES_REQUESTED** [review]
In this rebase again the cleanExpr PR by @Praveenrajmani is lost. will bring this back in.

**Comment by harshavardhana** [comment]
> In this rebase again the cleanExpr PR by @Praveenrajmani is lost. will bring this back in.

@sinhaashish can you bring the commit back?

**Comment by sinhaashish** [comment]
> > In this rebase again the cleanExpr PR by @Praveenrajmani is lost. will bring this back in.
> 
> @sinhaashish can you bring the commit back?

Done!

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#79218dae-2c7f-4232-b9ae-ecb9d2a33b8d) |
<a name="79218dae-2c7f-4232-b9ae-ecb9d2a33b8d"></a>
### 6648-8d9c0f3/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31523
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp fa3f6dbbd8d9:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 1 minutes and 22 seconds
(2/13) Running aws-sdk-go tests ... done in 4 seconds
(3/13) Running aws-sdk-java tests ... done in 10 seconds
(4/13) Running aws-sdk-php tests ... done in 1 minutes and 2 seconds
(5/13) Running aws-sdk-ruby tests ... done in 39 seconds
(6/13) Running mc tests ... done in 1 minutes and 1 seconds
(7/13) Running minio-dotnet tests ... done in 5 minutes and 38 seconds
(8/13) Running minio-go tests ... done in 2 minutes and 2 seconds
(9/13) Running minio-java tests ... FAILED in 2 minutes and 39 seconds
{
  "name": "minio-java",
  "function": "presignedGetObject(String bucketName, String objectName, Integer expires)",
  "duration": 71,
  "status": "FAIL",
  "error": "java.io.EOFException >>> [okio.RealBufferedSink.write(RealBufferedSink.java:107), io.minio.HttpRequestBody.writeTo(HttpRequestBody.java:74), okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:62), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:120), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:185), okhttp3.RealCall.execute(RealCall.java:69), io.minio.MinioClient.execute(MinioClient.java:1033), io.minio.MinioClient.executePut(MinioClient.java:1316), io.minio.MinioClient.executePut(MinioClient.java:1338), io.minio.MinioClient.putObject(MinioClient.java:3398), io.minio.MinioClient.putObject(MinioClient.java:3438), io.minio.MinioClient.putObject(MinioClient.java:3154), FunctionalTest.presignedGetObject_test2(FunctionalTest.java:1667), FunctionalTest.runTests(FunctionalTest.java:2729), FunctionalTest.main(FunctionalTest.java:2843)]"
}

Executed 8 out of 13 tests successfully.
```

File: pkg/s3select/input.go [hunk 1]
```diff
@@ -40,270 +36,97 @@ const (
...
-// stats represents a struct that represents the format for XML of the stat
-// messages
-type stats struct {
-	XMLName        xml.Name `xml:"Stats" json:"-"`
-	BytesScanned   int64    `xml:"BytesScanned"`
-	BytesProcessed int64    `xml:"BytesProcessed"`
-	BytesReturned  int64    `xml:"BytesReturned"`
+// convertMySQL Replaces double quote escape for column names with backtick for
+// the MySQL parser
+func convertMySQL(random string) string {
```
Comment (harshavardhana) [inline pkg/s3select/input.go]: This function should be removed with cleanExpr change @sinhaashish

Comment (sinhaashish) [reply pkg/s3select/input.go]: Modified

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by harshavardhana — APPROVED** [review]
LGTM we can merge and do more refactor on this change.

**Review by kannappanr — APPROVED** [review]
LGTM

**Pull request merged**